snuba/cogs/accountant.py

```diff
@@ -0,0 +1,36 @@
+from __future__ import annotations
+
+import logging
+
+from arroyo.backends.kafka.consumer import KafkaProducer
+from usageaccountant import UsageAccumulator, UsageUnit
+
+from snuba.utils.streams.configuration_builder import build_kafka_producer_configuration
+from snuba.utils.streams.topics import Topic as StreamTopic
+
+logger = logging.getLogger("usageaccountant")
+
+
+accumulator: UsageAccumulator | None = None
```

Comment:
I'll let the Storage team chime in on whether to use a module attribute or whether to tie the accumulator to the MessageProcessor lifecycle. 
Just remind that, as this contains a `KafkaProducer` that has a threadpool that has to be closed when the application terminates.

Reply:
cc @lynnagara, @untitaker may know more.

Reply:
I would also prefer it as part of the processor's lifecycle. Doing side-effects in MessageProcessor like this will conflict with the design of Rust consumers where we mostly assume that they are side-effect free transformation functions, but i don't really know right now if there is a better way to achieve this.

Reply:
Another way to do this would be to spin up an independent consumer on the gen-metrics topic which has the sole job of reporting this data using the UsageAccumulator. It would mean introducing a new piece of infra which just has the new consumer and it would be another consumer to maintain, but it would mean we don't have to touch the existing logic.

I'm not sure if we're ok with doing that (it would mean introducing a new consumer per shared resource consumer when accounting cogs), but if the conflict with the rust consumer design is too great it could be something to consider.

Reply:
let's just put it in the metrics processor for now. would prefer avoiding globals. if we have a need to move the generic metrics dataset to pure rust, we can create a separate consumer

Reply:
I would discourage adding a consumer for that. This library is supposed to be really lightweight.
Also having a separate consumer would be very fragile as you will have to keep the code in sync between the two if a pre-filter is added to the snuba main consumer that drops more events.

This is python, there is no issue in having side effects here as this is equivalent to recording metrics or logging. 

The alternative is to make a change to the Snuba consumer to add a place to record COGS like we do for recording end to end latency (that is recorded in the flush step).

Reply:
> This is python, there is no issue in having side effects here as this is equivalent to recording metrics or logging.

there is no concern with having the python logic do more, but should we at some point have a need to port the message processor to rust, we will have to port all of this logic to Rust, and we'd be rolling out our first rust producer (next to relay) at the same time.

i'm not thrilled to allow for (potentially async) I/O in message processors, but as said I think having it in the message processor is fine for now, it's just an architectural decision I think we should revisit in case we port to rust

Reply:
> The alternative is to make a change to the Snuba consumer to add a place to record COGS like we do for recording end to end latency (that is recorded in the flush step).

if this logic is supposed to be reusable across datasets then this would be preferrable

Reply:
One problem with the global variable approach. 
As all Arroyo producer. You need to close it before stopping the application or your application will hang till it gets a SIGTERM. 
(try to use the accumulator and flush it from a python repl then exit the repl without calling close(), it will hang there).

I don't see a call to `close()` in this approach. 
If you tie that to the lifecycle of the processor you can ensure it gets closed. 
If your consumer hang when closing you will have problems in production.

Reply:
The issue of a SIGTERM is that the Arroyo producer keeps a threadpool around. a threadpool needs to be manually closed before leaving the application.

Reply:
Updated the Consumer SIGINT/SIGTERM handler with a call to flush and close the cogs recorder (if one exists). The consumer no longer hangs when killed when killed.